### ISSUE TITLE
Parser confuses type-annotation expressions with pi types

### ISSUE THREAD

**Issue by danieldjohnson** [google-research/dex-lang#282]
The infix operator `(:)` can be used to annotate the types of subexpressions. However, this can lead to some ambiguities with parsing pi (dependent function) types when the expression being annotated is a single identifier. For instance, the expression
```
foo : Bar -> Baz
```
can be interpreted either as annotating the value `foo` to have type `Bar -> Baz`, or as a pi type binding `foo` to a value of type `Bar` and returning `Baz` (in this case, equivalent to the type `Bar -> Baz` itself).

Right now, the parser always interprets an identifier followed by `:` as a pi type, which sometimes leads to confusing error messages and isn't always the ideal behavior. This happens even when there is not an arrow afterward (which isn't actually ambiguous, but we commit early to parsing a pi type and throw an error).

One concrete instance where this comes up is when annotating polymorphic objects like `zero`:
```
:p zero : Float32
> Parse error:1:18:
>   |
> 1 | :p zero : Float32
>   |                  ^
> unexpected newline
> expecting ''', '_', alphanumeric character, or arrow
```
A workaround is to wrap the annotated subexpression in parentheses:
```
:p (zero) : Float32
> 0.0
```

My gut feeling is that it's OK to treat expressions like `foo : Bar -> Baz` as pi types, but that we should probably not be committing early to the pi-type hypothesis so that expressions like `zero : Float32` are well formed. But even having a clearer error message in this case would be an improvement.

Another possible fix would be to use some different delimiter for annotating pi type binders v.s. annotating the type of an expression.

**Comment by apaszke** [comment]
I think it might make sense to require the pi binders to be wrapped in parens. Our current parser actually rejects that, but it was the default that I've always been trying to use subconsciously... WDYT?

**Comment by danieldjohnson** [comment]
That seems reasonable to me, I always try to write parens too and then have to remove them. Although it's worth noting that doing that wouldn't remove the ambiguity, just choose how to resolve it (in a hopefully intuitive way). And we'd still need to be careful not to commit early, so that something like
```
(a : Int) -> Int
```
is a Pi type, but something like
```
(zero : Int) + 3
```
is an expression.

Thinking about this issue some more, I realized there's sort of an inconsistency around the "return value" of `(:)`. When used in a Pi type, `a:X -> Y` (or perhaps `(a:X) -> Y`) is equivalent to `X -> Y` (if `a` isn't in `Y`). But when used as a subexpression, `f (a:X)` is equivalent to `f a`. So depending on context, `:` either "returns" the right hand side or the left hand side.

As far as I can tell, it doesn't seem like Idris has an infix `:` for type annotations. I suppose we could define a special keyword other than `:` for type annotation expressions, for instance `a as X`, which might remove that inconsistency.

**Comment by apaszke** [comment]
Good analysis! I agree that using a different operator might make sense, although it does make me a little sad, because in both the cases you've identified it really is a type annotation.

I guess that the big question is whether you read `(a : Int)` as a reference to an in-scope variable `a` and additionally constrain its type to be `Int`, or if you treat it as a binder. They're just completely different syntactic entities, but distinguishing them seems to require a pretty huge lookahead.

**Comment by axch** [comment]
As written, this was fixed by #938, though we still have a parse ambiguity of sorts between pi types and type-annotated let binders.
